File: problem2.py
# Name: problem2.py
# Author: Austin Vargason
# Description: Implementing problem 2 Haskell code in Python

from typing import TypeVar, List, Callable, Any, Iterable
from pyrsistent import pvector
from toolz import flip, compose, curry, concat
from functools import wraps
from itertools import cycle
import operator

#type variables
A = TypeVar('A')
B = TypeVar('B')
C = TypeVar('C')

#necessary prereq functions
# zipWith function in python
# zipWith takes a function and 2 lists and zips the list using the operation provided
@curry
def myZipWith(f: Callable[[Any, Any], Any], listA: List[Any], listB: List[Any]) -> List[Any]:
    """Return two lists added together
    >>> listA = [1, 2, 3, 4, 5]
    >>> listB = [6, 7, 8, 9, 10]
    >>> myZipWith(operator.add, listA, listB)
    [7, 9, 11, 13, 15]
    """
    return [(f(x[0],x[1])) for x in zip(listA, listB)]
    #alternative
    #return list(map(f, zip(listA, listB)))

#
def reverse(values: List[int]) -> List[int]:
    return values[::-1]

# toDigit function
def toDigit(digit: str) -> int:
    return int(digit)

# to Digits function
@curry
def toDigits(n: int) -> Iterable[int]:
    return pvector(map(toDigit, str(n)))

# doubleEveryOther function
@curry
def doubleEveryOther(values: Iterable[int]) -> Iterable[int]:
    return compose(myZipWith(operator.mul, cycle([1, 2])), reverse)(values)

# sumDigits function
@curry
def sumDigits(values: Iterable[int]) -> int:
    curryMap = curry(map)
    return compose(sum, concat, curryMap(toDigits))(values)

#checkSum function
@curry
def checkSum(n: int) -> int:
    return compose(sumDigits, doubleEveryOther, toDigits)(n)

#isValid function
@curry
def isValid(n: int) -> bool:
    curryEquals = curry(operator.eq)
    curryMod = curry(operator.mod)
    return compose(curryEquals(0), flip(curryMod, 10), checkSum)(n)

# myFlip function
def myFlip(f: Callable[[A, B], C]) -> Callable[[B, A], C]:
    @wraps(f)
    def newF(*args):
        return f(*args[::-1])
    return newF

# testCC function
def testCC() -> Iterable[bool]:
    return list(map(isValid, [1234567890123456, 1234567890123452]))


if __name__ == "__main__":
    print(testCC())

File: problem2.hs
-- Name: problem2.hs
-- Description: Commented Haskell Code from assignment document
-- Author: Austin Vargason

-- many of these functions are written in point-free notation
-- point free style is when you define functions as a composition
-- of functions where we never mention the actual arguments the function may be applied to
-- point-free style is considered cleaner and more compact

-- toDigit function: In this function we take in a character c as an argument and provide it as an arg to the read function
-- which takes in a string, we can convert c to a string by making it a single element array containing c
-- the read function will parse the character c into a integer digit
toDigit :: Char -> Int
toDigit c = read [c]

-- toDigits function: this function takes in an Int and produces and Int array.
-- In this function we use a function composition to take the result of the show function on the Int to convert to a string
-- and use the map function to map toDigit to each char in the string
toDigits :: Int -> [Int]
toDigits = map toDigit . show

-- doubleEveryOther function: this function takes in an Int array and produces an Int array
-- this function uses function composition to reverse the Int array and passes the result to the zipWith function that multiplies every number by the number in the list created by cycle (either 1 or 2)
-- this will double every other number in the array
doubleEveryOther :: [Int] -> [Int]
doubleEveryOther = zipWith (*) (cycle [1,2]) . reverse

-- sumDigits function: this function takes in an Int array and produces an Int result
-- this function uses function composition to take the result of mapping an Int array to the toDigits function, then concattenating the lists together in a single list, then summing the elements in that list
sumDigits :: [Int] -> Int
sumDigits = sum . concat . map toDigits

-- checkSum function: takes in an Int and produces an Int
-- this function uses function composition to take the result of ToDigits called on the Int provided and pass it to the doubleEveryOther function, which then gets passes to the sumDigits function to produce a single sum
checkSum :: Int -> Int
checkSum = sumDigits . doubleEveryOther . toDigits

-- isValid function: this function takes an Int and produces a Bool
-- this function uses function composition to pass the result of the checksum function called on the Int provided to the (flip mod) 10 call (flip flips the argument order for the function), this takes the the mod 10 of the result,
-- then passes it to the (==) function to check whether the result is equal to 0.
-- overall this function does a checksum on the Int and checks whether that number is equally divisible by 10 to determine a valid result
isValid :: Int -> Bool
-- This simple version:
-- isValid n = checkSum n `mod` 10 == 0
-- can be tortured into this point-free form:
isValid = (==) 0 . (flip mod) 10 . checkSum
-- Can you explain the point-free form?
-- (This is an exercise, not a claim that it’s better code.)

-- flip is defined in the Prelude, but it can be defined as follows.
myFlip :: (a -> b -> c) -> (b -> a -> c)
myFlip f = \x y -> f y x

-- testCC function: this function produces an array of Bools
-- this function maps the isValid function to two credit card numbers to determine whether they are valid
testCC :: [Bool]
testCC = map isValid [1234567890123456, 1234567890123452]
-- => [False, True]